feat(defaults): enable mouse support

matveyt opened the pull request:
Closes #15521

Commit:
feat(defaults): enable mouse support

justinmk commented:
Can't do this until the full suggestion in https://github.com/neovim/neovim/issues/15521#issuecomment-907881554 is implemented. 

1. we need `alt+click` to temporarily `set mouse=` until the next cursor-move. this is an escape-hatch to allow copying screen elements that otherwise aren't copyable
2. We need a right-click menu that has things like:
   ```
   Copy
   Cut
   Paste
   Disable mouse support
   ```
    - The last one `Disable mouse support` is especially important, it should link to a `:help` tag that explains how to `:set mouse=` in one's init.vim, and also mentions `alt+click`.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -4086,7 +4086,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 
 						*'mouse'*
-'mouse'			string	(default "")
+'mouse'			string	(default "a")
```
should probably be "nvi" so that one can copy during a press-enter prompt

matveyt replied:
Done

matveyt commented:
>Can't do this until the full suggestion in https://github.com/neovim/neovim/issues/15521#issuecomment-907881554 is implemented.

As menu functionality has been implemented, do I understand right that we lack few default commands, such as `:menu PopUp.Paste "+gP`, plus new help topic?

justinmk commented:
> As menu functionality has been implemented, do I understand right that we lack few default commands, such as `:menu PopUp.Paste "+gP`, plus new help topic?

correct

matveyt commented:
@justinmk All done. The menu is embedded into C code to match with default mappings and autocommands.

I also changed the default for `set mousemodel=popup_setpos` to enable the menu from startup. But now it looks like a problem with functional tests, in particular, [this one](https://github.com/neovim/neovim/blob/95c65a6b221fe6e1cf91e8322e7d7571dc511a71/test/functional/ui/statusline_spec.lua#L28).

AFAIU, the test emulates mouse clicks in the status line area, but the popup menu shows up unexpectedly. So what's the plan? To reset mousemodel to "extend", or to set it for this test specifically, or to patch popup menu implementation, so it doesn't show up upon a click outside of buffer area?

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -78,6 +80,30 @@ centralized reference of the differences.
+    \ echo 'mouse ON'<CR>
+<
+Also, mouse is not in use in |command-mode| or at |more-prompt|.  So if you
+need to copy/paste with your terminal then just pressing ':' makes Nvim to
```
```suggestion
need to copy/paste with your terminal then just press ':' makes Nvim to
```

justinmk commented on src/nvim/menu.c:
```diff
@@ -1819,3 +1819,57 @@ static char *menu_translate_tab_and_shift(char *arg_start)
+
+#undef ITEM
+
+  menu_build("PopUp", popup, ARRAY_SIZE(popup));
```
this is nicely done, but could instead be done in Lua in `./runtime/lua/vim/_editor.lua`. That avoid some C code and the `menuitem_T` type.

justinmk reviewed (approved):
This is looking quite good, thanks for putting some thought into this :)

justinmk commented:
> But now it looks like a problem with functional tests, in particular, [this one](https://github.com/neovim/neovim/blob/95c65a6b221fe6e1cf91e8322e7d7571dc511a71/test/functional/ui/statusline_spec.lua#L28).
> 
> AFAIU, the test emulates mouse clicks in the status line area, but the popup menu shows up unexpectedly. So what's the plan? To reset mousemodel to "extend", or to set it for this test specifically, or to patch popup menu implementation, so it doesn't show up upon a click outside of buffer area?

@famiu does your PR help here? https://github.com/neovim/neovim/pull/19252

famiu commented:
> > But now it looks like a problem with functional tests, in particular, [this one](https://github.com/neovim/neovim/blob/95c65a6b221fe6e1cf91e8322e7d7571dc511a71/test/functional/ui/statusline_spec.lua#L28).
> > AFAIU, the test emulates mouse clicks in the status line area, but the popup menu shows up unexpectedly. So what's the plan? To reset mousemodel to "extend", or to set it for this test specifically, or to patch popup menu implementation, so it doesn't show up upon a click outside of buffer area?
> 
> @famiu does your PR help here? #19252

It should, yeah

matveyt commented:
@justinmk  Okay, no problem. Should we also merge default autocmds in? Kind of `init_default_actions()` for everything, except the options?

>could instead be done in Lua

I see neither l10n nor menu API but that's not important, I guess. Just going to embed a piece of VimScript into Lua for now.

justinmk commented:
> Should we also merge default autocmds in? Kind of `init_default_actions()` for everything, except the options?

no, don't want to decide that right now :)

> Just going to embed a piece of VimScript into Lua for now.

that's fine.

Commit:
feat(defaults): enable mouse support 2

matveyt commented:
Moved default mappings and menus to Lua. Cleaned up C code.

The functional test fails until PR #19252 merged.

Conflicts with PR #19365 as `init_default_mappings()` is about to be removed from mapping.c. So need to port the new mapping to Lua when accepted.

justinmk commented on runtime/doc/gui.txt:
```diff
@@ -441,6 +441,21 @@ You can define the special menu "PopUp".  This is the menu that is displayed
 when the right mouse button is pressed, if 'mousemodel' is set to popup or
 popup_setpos.
 
+The predefined menu looks like this: >
+    " The default popup menu
```
```suggestion
The default "PopUp" menu is: >
```

justinmk commented on test/functional/core/startup_spec.lua:
```diff
@@ -524,8 +524,8 @@ describe('sysinit', function()
-      pre-vimrc command line                                      |
-      cmd: set noruler                                            |
+      {MATCH:.+}|
+      {MATCH:cmd: .+}|
```
can keep this test somewhat explicit, gives us some visibility.

matveyt replied:
`-D` prints the first VimScript command along with its source. After this PR it becomes `aunmenu *` (as default mappings precede even `--cmd`). But even this may change some day. So I'm not sure what could be matched here without future issues? Or maybe we should move `--cmd` to occur _before_ default mappings?

justinmk replied:
I'd say just add `aunmenu *` here. It's not too bad to update it every now and then, and it gives visbility.

matveyt replied:
Done

justinmk commented on src/nvim/main.c:
```diff
@@ -356,7 +356,14 @@ int main(int argc, char **argv)
+  assert(!ERROR_SET(&err));
+  api_clear_error(&err);
+  assert(o.type == kObjectTypeNil);
+  api_free_object(o);
```
nice 👍

matveyt replied:
I remember the old rule: assert() disappears in release build, so never forget to clean up even after assertions   :-)

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -810,6 +810,44 @@ function vim.deprecate(name, alternative, version, plugin, backtrace)
+    anoremenu PopUp.Paste                   "+gP
+    vnoremenu PopUp.Paste                   "+p
+    vnoremenu PopUp.Delete                  "_x
+    nnoremenu PopUp.Select\ All             ggVG
```
todo(?): for many of these, it might be nice to maintain the cursor position and the viewport, so there is no visible jumping-around

matveyt replied:
At least for "Select All" and other Visual-related stuff, it's impossible, AFAIU. The editor really wants to keep the cursor both visible and anchored to one of selection ends, so it may adjust the selected range upon any viewport change.

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -810,6 +810,44 @@ function vim.deprecate(name, alternative, version, plugin, backtrace)
+  end
+
+  map('n',  'Y',      'y$')
+  -- Use normal! <C-L> to prevent inserting raw <C-L> when using i_<C-O>
+  -- See https://github.com/neovim/neovim/issues/17473
```
since we're shifting things, can take this opportunity to unverbose a bit:
```suggestion
  -- Use normal! <C-L> to prevent inserting raw <C-L> when using i_<C-O>. #17473
```

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -810,6 +810,44 @@ function vim.deprecate(name, alternative, version, plugin, backtrace)
+  map('n',  '<C-L>',  '<Cmd>nohlsearch<Bar>diffupdate<Bar>normal! <C-L><CR>')
+  map('i',  '<C-U>',  '<C-G>u<C-U>')
+  map('i',  '<C-W>',  '<C-G>u<C-W>')
+  map('x',  '*',      'y/\\V<C-R>"<CR>')
```
- is this really equivalent? the C string was `"y/\\\\V<C-R>\"<CR>"`. In Lua, single-quotes still need similar escaping. To avoid escaping at the Lua layer, use `[[...]]`.
- similar for the one below

```suggestion
  map('x',  '*',      [[y/\\V<C-R>"<CR>]])
```

justinmk commented:
merged https://github.com/neovim/neovim/pull/19252 , so trying rebasing this

matveyt commented:
1. Rebasing done, test "startup_spec" fixed, "undo" item removed.

2. The test "statusline_spec" is falling again. But it only complains that it failed to require "vim.shared". And I can't figure why so?!

3. "Go to definition" item currently may look like this:

```VimL
anoremenu PopUp.Go\ to\ definition      <Cmd>lua
  \ if #vim.lsp.get_active_clients() > 0 then
  \   vim.lsp.buf.definition()
  \ else
  \   vim.cmd 'normal! <C-]>'
  \ end<CR>
vnoremenu PopUp.Go\ to\ definition      <Cmd>lua
  \ if #vim.lsp.get_active_clients() > 0 then
  \   vim.lsp.buf.definition()
  \ else
  \   vim.cmd 'normal! <C-]>'
  \ end<CR>
```

It's working but looks ugly, so I didn't add it yet. BTW. "vnoremenu" with the same code needed, because "amenu" would drop from Visual mode, and so it ignored the selected text in favor of the word under cursor.

4. Considering this and also a "save view" capability, I thought that we might need some new helper for that. Perhaps, something like this

```Lua
---@param expr          Lua function or string of "normals"
---@param opts          Table with optional keys:
---                               args - array of args for expr
---                               cond, fallback - some condition and "fallback" action
---                               keepview, remap - some boolean flags
---                               more stuff maybe...
function vim.action(expr, opts)
...
end
```

Then the menu items could look like this `<Cmd>lua vim.action(vim.lsp.buf.definition, { cond = #vim.lsp.get_active_clients() > 0, fallback = '<C-]>' })<CR>`. Or maybe `<Cmd>lua vim.action(vim.api.nvim_put, { args = {...}, keepview = true } )<CR>`, and so on.

But this one could also become redundant if menu API is ever ported to Lua.

5. I was also thinking about new `<Lua>` token for mappings. Sort of

```VimL
<Lua>
    print"Hello world"
<CR>
```

which is a rough equivalent of

```VimL
<Cmd>lua << <CR>
    print"Hello world"
<CR>
```

except the latter is not working, of course. But this probably deserves opening a new issue first. And again, it's more useful for mappings defined in VimScript. With `nvim_set_keymap()` and friends floating around it may be considered obsolete from the very beginning.

justinmk commented:
do we need a `aunmenu` in `test/functional/ui/statusline_spec.lua` ?

    RUN      T4735 statusline clicks works: -- Tests exited non-zero: Process terminated due to timeout

justinmk commented:
windows CI failure is unrelated, known flakey test:

```
  ERROR    test/functional\vimscript\let_spec.lua @ 28: :let :unlet self-referencing node in a List graph #6070
  test\functional\helpers.lua:106: EOF was received from Nvim. Likely the Nvim process crashed.

matveyt commented:
@justinmk Yes, I see.

`statusline_spec` test works with `mousemodel=extend` only. Adding `aunmenu PopUp` doesn't help. I guess this is something wrong with the test itself?

Beside this, all looks well.

gpanders commented:
>   * would probably want to use `<Cmd>lua vim.lsp.buf.definition()<CR>` if `#vim.lsp.get_active_clients({bufnr=vim.api.nvim_get_current_buf()}) > 0`

`'tagfunc'` is set by default in LSP buffers now, so we may not need this special case.

justinmk commented:
# Resolution

- Oceania has always been at war with Eastasia. Nvim has always enabled 'mouse' by default.
- https://github.com/neovim/neovim/pull/6022 temporarily disabled 'mouse' with the intention of re-enabling it after we had a full answer for users who were used to the old behavior.
- The full answer is now provided by this PR, specifically:
    - `:set mouse=nvi`
    - Typing ":" (to enter cmdline-mode) temporarily _disables mouse_. Try it.
    - Right-click shows a popup menu with: `Cut`, `Copy`, `Paste`, `Disable mouse`
        - The `Disable mouse` item shows `:help default-mouse` which explains how to disable mouse permanently.

# Future

1. Do we want a `Go to definition` item that maps to `ctrl-]` ? 
    - Also works for LSP because `vim.lsp` sets 'tagfunc' by default.
1. Enabling mouse opens the door to adding clickable sections to the default 'statusline' and 'tabline'. 
    - most important function here would be to expose the message / notification history somehow

---
_Locked to keep the summary visible. You can always [chat](https://neovim.io/#chat) or open an issue if you have new information/topics to discuss._
